### PR TITLE
fix(firehose): build S3 object keys with AWS prefix expressions and name suffix

### LINKED ISSUE #1857: [BUG] Firehose: S3 delivery object keys don't match AWS (prefix semantics and object name format)
## Problem

When Firehose delivers buffered records to S3, Floci builds the object key in a way that doesn't match AWS:

| Aspect | Floci today | Real AWS |
|---|---|---|
| Default prefix (no `Prefix`) | `<streamName>/` | `yyyy/MM/dd/HH/` (UTC) |
| Static `Prefix` | used as-is, trailing `/` forced | literal prefix + `yyyy/MM/dd/HH/` appended (no `/` inserted) |
| Prefix expressions | non-AWS placeholders `{year}`, `{month}`, `{day}`, `{hour}` | `!{timestamp:<DateTimeFormatter pattern>}`, `!{firehose:random-string}` |
| Object name | `<uuid>.json` | `<streamName>-<versionId>-<yyyy-MM-dd-HH-mm-ss>-<uuid>` (no extension when UNCOMPRESSED) |
| `CustomTimeZone` | ignored | timezone used for prefix/suffix timestamps (default UTC) |

Current implementation: `FirehoseService.resolvePrefix()` and the key construction in `flush()` (`src/main/java/io/github/hectorvent/floci/services/firehose/FirehoseService.java`).

This breaks consumers that rely on AWS's partitioned layout (e.g. Glue/Athena tables partitioned by the default `yyyy/MM/dd/HH/` hierarchy, or tools that parse the object-name suffix).

## Evidence — observed on real AWS (2026-07-13, eu-west-1)

DirectPut streams with an ExtendedS3 destination and zero buffering; one record put per stream; keys captured with `list-objects-v2`:

| `Prefix` configured | Observed S3 key |
|---|---|
| *(unspecified)* | `2026/07/13/14/floci-compat-default-1-2026-07-13-14-24-41-d1bc712a-d08c-4d8c-9f47-84445d891036` |
| `events/data/` | `events/data/2026/07/13/14/floci-compat-static-1-2026-07-13-14-24-42-9088be6e-1fec-4048-85bf-fd6d77ae6843` |
| `legacy` (no trailing slash) | `legacy2026/07/13/14/floci-compat-noslash-1-2026-07-13-14-24-43-665cbd00-f6f7-444f-8002-691a1c41abbb` |
| `data/!{timestamp:yyyy/MM/dd}/` | `data/2026/07/13/floci-compat-expr-1-2026-07-13-14-24-44-2f17ec37-432b-4194-beb4-50496ee733c1` |
| `rand/!{firehose:random-string}/` | `rand/84GCfdydsgd/floci-compat-rand-1-2026-07-13-14-24-45-b84fccd9-efa5-4828-9206-f9f371567c91` |

Notes beyond what the docs say:

- The `yyyy/MM/dd/HH/` append happens only when the prefix contains **no `!{...}` expression at all**. The [docs' semantic rules](https://docs.aws.amazon.com/firehose/latest/dev/s3-prefixes.html) claim the append happens whenever there is no *timestamp* expression, but a prefix whose only expression is `!{firehose:random-string}` did **not** get the time prefix appended.
- `!{firehose:random-string}` evaluates to an 11-char mixed-case alphanumeric string (`84GCfdydsgd`), not the UUID-like hex shown in the docs example.
- CreateDeliveryStream validation messages (for reference; not necessarily in scope):
  - Prefix with expressions + null `ErrorOutputPrefix` → `InvalidArgumentException: The supplied prefix(es) do not satisfy the following constraint: ErrorOutputPrefix cannot be null or empty when Prefix contains expressions`
  - `!{bogus:xx}` → `InvalidArgumentException: The supplied prefix(es) do not satisfy the following constraint: Namespace to the left of colon (bogus) is an invalid keyword!`

References: [Custom prefixes for Amazon S3 objects](https://docs.aws.amazon.com/firehose/latest/dev/s3-prefixes.html), [S3 object name format](https://docs.aws.amazon.com/firehose/latest/dev/s3-object-name.html).

## Proposed fix

- Evaluate `!{timestamp:<pattern>}` (Java `DateTimeFormatter`, all instances sharing one instant) and `!{firehose:random-string}` (fresh 11-char alphanumeric per instance) in `Prefix`.
- Append `yyyy/MM/dd/HH/` when the prefix has no expressions (including null/empty prefix), by literal concatenation.
- Name objects `<streamName>-<versionId>-<yyyy-MM-dd-HH-mm-ss>-<uuid>` with no extension.
- Support `CustomTimeZone` (default UTC).
- Unsupported namespaces (`partitionKeyFromQuery`/`partitionKeyFromLambda`/unknown) would be left literally in the key (Floci doesn't implement dynamic partitioning), and timestamps would be evaluated at flush time rather than at oldest-record arrival — both documented as known deviations.

I have a patch for this and can open a PR.

### PR DESCRIPTION
**Pull request by mihaiblidaru**
Fixes #1857

## What

Firehose S3 deliveries now build object keys the way AWS does: `<evaluated prefix><streamName>-<versionId>-<yyyy-MM-dd-HH-mm-ss>-<uuid>` (no file extension), replacing the previous `<prefix-or-streamName/><uuid>.json`.

- No `Prefix` → default prefix `yyyy/MM/dd/HH/` (UTC).
- A `Prefix` with no `!{...}` expression gets `yyyy/MM/dd/HH/` appended by literal concatenation — no `/` inserted (`legacy` → `legacy2026/07/13/14/...`), exactly as observed on real AWS.
- `!{timestamp:<pattern>}` (Java `DateTimeFormatter`, single-quoted literals honored, all instances share one instant) and `!{firehose:random-string}` (fresh 11-char alphanumeric per instance) are evaluated; a prefix containing any expression gets nothing appended.
- New `CustomTimeZone` support on the S3 destination (create/update/describe + key evaluation); absent or invalid zones fall back to UTC.
- The non-AWS `{year}/{month}/{day}/{hour}` placeholders are removed (in AWS they are literal text).

Implementation: new package-private `S3ObjectKeyResolver` (regex-driven expression scanner, never throws — malformed/unsupported expressions are kept literally and logged, so a flush never drops data), `FirehoseService` now injects `Clock` (ClockProducer/MutableClock pattern) and delegates key construction to the resolver.

## Evidence

All rules were captured empirically against a real AWS account (eu-west-1, zero-buffering streams) — full table and validation-error messages in #1857. Two notable findings beyond the docs:

- The docs say the default time prefix is appended whenever there is no *timestamp* expression, but on real AWS **any** expression suppresses the append (a `rand/!{firehose:random-string}/` prefix got nothing appended).
- `!{firehose:random-string}` yields mixed-case alphanumerics (e.g. `84GCfdydsgd`), not the hex shown in the docs example.

Replaying the same five prefix scenarios against this patch produces structurally identical keys to real AWS (comparison table in #1857).

## Known deviations (documented in docs/services/firehose.md)

- Timestamps are evaluated at flush time rather than the oldest buffered record's arrival time.
- Expressions AWS rejects at create time (unknown namespaces, dynamic-partitioning keys, invalid patterns) are kept literally in the key; no create/update-time prefix validation was added.
- Content is never compressed, so no compression extension is added.

## Tests

- New `S3ObjectKeyResolverTest` (17 tests, one per evaluation rule/edge case) and `FirehoseServiceTest` (4 tests, MutableClock + mocked S3Service).
- `FirehoseExtendedS3IntegrationTest` extended with 6 delivery-plane tests (one per prefix scenario, asserting the delivered key via the S3 ListBucket XML).
- Tests that relied on the old emulator-specific `<streamName>/` default prefix now declare it explicitly: `SesEventPublishingV2IntegrationTest`, compat `SesEventPublishingTest`, compat `DataLakeTest` (its Glue table keeps working — keys stay under the table location and the Athena engine reads it recursively).
- Full suite green (7821 tests; one unrelated Neptune socket-timeout flake passes in isolation). Compatibility suites run via the CI Docker method: `sdk-test-java` (1401 tests, only an unrelated AppSync 409 flake that passes in isolation, `DataLakeTest` included), `compat-terraform` 22/22, `compat-opentofu` 16/16.

### CODE REVIEW

File: src/main/java/io/github/hectorvent/floci/services/firehose/S3ObjectKeyResolver.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ * reject at create time (unknown namespaces, dynamic-partitioning keys,
+ * invalid patterns) are kept literally in the key instead of failing.
+ */
+final class S3ObjectKeyResolver {
+
+    private static final Logger LOG = Logger.getLogger(S3ObjectKeyResolver.class);
+    private static final Pattern EXPRESSION = Pattern.compile("!\\{(?<namespace>[^:}]*)(?::(?<argument>[^}]*))?}");
+    private static final String DEFAULT_TIMESTAMP_PATTERN = "yyyy/MM/dd/HH/";
+    private static final DateTimeFormatter SUFFIX_TIMESTAMP =
+            DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss", Locale.ROOT);
```
[the remark this answers is not in the document]
Comment (mihaiblidaru) [reply src/main/java/io/github/hectorvent/floci/services/firehose/S3ObjectKeyResolver.java]: Applied in 62e02548.

File: src/main/java/io/github/hectorvent/floci/services/firehose/S3ObjectKeyResolver.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        boolean expressionSeen = false;
+        Matcher matcher = EXPRESSION.matcher(prefix == null ? "" : prefix);
+        while (matcher.find()) {
+            expressionSeen = true;
+            matcher.appendReplacement(out, Matcher.quoteReplacement(evaluateExpression(matcher, time)));
+        }
+        matcher.appendTail(out);
+        if (!expressionSeen) {
+            out.append(DateTimeFormatter.ofPattern(DEFAULT_TIMESTAMP_PATTERN, Locale.ROOT).format(time));
+        }
```
[the remark this answers is not in the document]
Comment (mihaiblidaru) [reply src/main/java/io/github/hectorvent/floci/services/firehose/S3ObjectKeyResolver.java:63]: Applied in 62e02548.

**Review by pgermosen — APPROVED** [review]
Looks good.

Excellent PR — the empirical validation against real AWS (#1857), the never-throws resolver design, and the test coverage are all top-notch. `S3ObjectKeyResolver` faithfully reproduces the documented behavior: default `yyyy/MM/dd/HH/` appended by literal concatenation only when the prefix has no expression, `!{timestamp:...}` / `!{firehose:random-string}` evaluation, a single shared instant across all expressions, and the `<stream>-<version>-<ts>-<uuid>` suffix.
